catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/MultiTableUpdate.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2024 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.catalog.service.impl;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.projectnessie.catalog.model.snapshot.NessieEntitySnapshot;
+import org.projectnessie.client.api.CommitMultipleOperationsBuilder;
+import org.projectnessie.error.NessieConflictException;
+import org.projectnessie.error.NessieNotFoundException;
+import org.projectnessie.model.Branch;
+import org.projectnessie.model.CommitResponse;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.Operation;
+
+/** Maintains state across all individual updates of a commit. */
+final class MultiTableUpdate {
+  private final CommitMultipleOperationsBuilder nessieCommit;
+  private final List<SingleTableUpdate> tableUpdates = new ArrayList<>();
+  private final List<String> storedLocations = new ArrayList<>();
+  private Map<ContentKey, String> addedContentsMap;
+  private Branch targetBranch;
+  private boolean committed;
+
+  MultiTableUpdate(CommitMultipleOperationsBuilder nessieCommit, Branch target) {
+    this.nessieCommit = nessieCommit;
+    this.targetBranch = target;
+  }
+
+  MultiTableUpdate commit() throws NessieConflictException, NessieNotFoundException {
+    synchronized (this) {
```

Comment:
The synchronized block was there already but I was wondering if we really need it?

Reply:
Not really sure, but it might happen that different threads access the fiels (completion-stages)